### ISSUE TITLE
03 - Model v2 does not learn properly

### ISSUE THREAD

**Issue by oabaig** [mrdbourke/pytorch-deep-learning#1168]
I keep getting a train accuracy of 10.00% and a test accuracy of 9.99%. I have tried the course google colab version and I get the exact same. Not sure what exactly is going on, but either something changed with the model or in PyTorch that could have impacted the results.

Here is my current model v2, 

```
class FashionMNISTModelV2(nn.Module):
    def __init__(self, input_shape: int, hidden_units: int, output_shape: int):
        super().__init__()
        self.conv_block_1 = nn.Sequential(
            nn.Conv2d(in_channels=input_shape, out_channels=hidden_units, kernel_size=3, stride=1, padding=1),
            nn.ReLU(),
            nn.Conv2d(in_channels=hidden_units, out_channels=hidden_units, kernel_size=3, stride=1, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)
        )
        self.conv_block_2 = nn.Sequential(
            nn.Conv2d(in_channels=hidden_units, out_channels=hidden_units, kernel_size=3, stride=1, padding=1),
            nn.ReLU(),
            nn.Conv2d(in_channels=hidden_units, out_channels=hidden_units, kernel_size=3, stride=1, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2, stride=2)
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            nn.Linear(in_features=hidden_units*7*7, out_features=output_shape)
        )
    
    def forward(self, x):
        x = self.conv_block_1(x)
        # print(f"Output shape of conv block 1 {x.shape}")
        x = self.conv_block_2(x)
        # print(f"Output shape of conv block 2 {x.shape}")
        x = self.classifier(x)
        # print(f"Output shape of classifier {x.shape}")
        return x
 ```

Just to reiterate, I have tried the google colab notebook created by @mrdbourke and it gives me the same result.

**Comment by oabaig** [comment]
Mistakenly closed this issue because I thought I saw an issue, but the issue still seems to be persisting.

This is the result of running the block 7.4 in `03_pytorch_computer_vision.ipynb`:

![image](https://github.com/user-attachments/assets/03c8a581-e592-43ff-a1d8-31aea825c685)

**Comment by technOslerphile** [comment]
I am also getting the exact same results as @oabaig . Something is weird. Could be PyTorch version differences? Unable to figure it out yet...

![Image](https://github.com/user-attachments/assets/3a3eaad7-30e6-41cf-bb0b-be1a5a7b7dff)

**Comment by technOslerphile** [comment]
I just solved this issue by doing this - I instantiated my loss function and optimizer AFTER instantiating the model_2 class. Due to so much code, there were two model_2 instances in my code, one before the loss and optimizer instances and one after. 
```class FashionMNISTModelV2(nn.Module):
    def __init__(self, input_shape: int, hidden_units: int, output_shape: int):
        super().__init__()
        self.block_1 = nn.Sequential(
            nn.Conv2d(in_channels=input_shape,
                      out_channels=hidden_units,
                      kernel_size=3, # how big is the square that's going over the image?
                      stride=1, # default
                      padding=1),# options = "valid" (no padding) or "same" (output has same shape as input) or int for specific number
            nn.ReLU(),
            nn.Conv2d(in_channels=hidden_units,
                      out_channels=hidden_units,
                      kernel_size=3,
                      stride=1,
                      padding=1),
            nn.ReLU(),
            nn.MaxPool2d(kernel_size=2,
                         stride=2) # default stride value is same as kernel_size
        )
        self.block_2 = nn.Sequential(
            nn.Conv2d(hidden_units, hidden_units, 3, padding=1),
            nn.ReLU(),
            nn.Conv2d(hidden_units, hidden_units, 3, padding=1),
            nn.ReLU(),
            nn.MaxPool2d(2)
        )
        self.classifier = nn.Sequential(
            nn.Flatten(),
            # Where did this in_features shape come from?
            # It's because each layer of our network compresses and changes the shape of our input data.
            nn.Linear(in_features=hidden_units*7*7,
                      out_features=output_shape)
        )

    def forward(self, x: torch.Tensor):
        x = self.block_1(x)
        #print(x.shape)
        x = self.block_2(x)
        #print(x.shape)
        x = self.classifier(x)
        #print(x.shape)
        return x

torch.manual_seed(42)
model_2 = FashionMNISTModelV2(input_shape=1, #number of color channels in input channel
    hidden_units=10,
    output_shape=len(class_names)).to(device)
model_2
# Setup loss and optimizer
loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(params=model_2.parameters(),
                             lr=0.1)
```

```
torch.manual_seed(42)

# Measure time
from timeit import default_timer as timer
train_time_start_model_2 = timer()

# Train and test model
epochs = 3
for epoch in tqdm(range(epochs)):
    print(f"Epoch: {epoch}\n---------")
    train_step(data_loader=train_dataloader,
        model=model_2,
        loss_fn=loss_fn,
        optimizer=optimizer,
        accuracy_fn=accuracy_fn,
        device=device
    )
    test_step(data_loader=test_dataloader,
        model=model_2,
        loss_fn=loss_fn,
        accuracy_fn=accuracy_fn,
        device=device
    )

train_time_end_model_2 = timer()
total_train_time_model_2 = print_train_time(start=train_time_start_model_2,
                                           end=train_time_end_model_2,
                                           device=device)
```

And the output is here which looks alright to me...

![Image](https://github.com/user-attachments/assets/9d4e7195-784b-4163-85ac-e8a1068ab6af)

**Comment by oabaig** [comment]
Great, I'll try this out when I get the chance!

**Comment by oabaig** [comment]
Can confirm that this works on my end
